beacon-chain/sync/data_column_sidecars.go

```diff
@@ -39,249 +39,394 @@ type DataColumnSidecarsParams struct {
+		allStoredColumns, err := storage.Get(root, nil)
 		if err != nil {
-			return nil, errors.Wrapf(err, "try get direct columns for root %#x", root)
+			return nil, errors.Wrapf(err, "failed to get all stored columns for reconstruction for block root %#x", root)
 		}
 
-		result[root] = append(result[root], storedColumns...)
+		// Attempt reconstruction.
+		reconstructedColumns, err := peerdas.ReconstructDataColumnSidecars(allStoredColumns)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to reconstruct data columns for block root %#x", root)
+		}
+
+		// Health check: ensure we have the expected number of columns.
+		numberOfColumns := params.BeaconConfig().NumberOfColumns
+		if uint64(len(reconstructedColumns)) != numberOfColumns {
+			return nil, errors.Errorf("reconstructed %d columns but expected %d for block root %#x", len(reconstructedColumns), numberOfColumns, root)
+		}
+
+		// Extract only the requested indices from reconstructed data using direct indexing.
+		for _, index := range requestedIndices {
+			if index >= numberOfColumns {
+				return nil, errors.Errorf("requested column index %d exceeds maximum %d for block root %#x", index, numberOfColumns-1, root)
+			}
+			result[root] = append(result[root], reconstructedColumns[index])
+		}
+
+		delete(roots, root)
 	}
 
 	return result, nil
 }
 
-// tryGetStoredColumns attempts to retrieve all requested data column sidecars directly from storage
-// if they are all available. Returns the sidecars if successful, and nil if at least one
-// requested sidecar is not available in the storage.
-func tryGetStoredColumns(storage filesystem.DataColumnStorageReader, blockRoot [fieldparams.RootLength]byte, indices []uint64) ([]blocks.VerifiedRODataColumn, error) {
-	// Check if all requested indices are present in cache
-	storedIndices := storage.Summary(blockRoot).Stored()
-	allRequestedPresent := true
-	for _, requestedIndex := range indices {
-		if !storedIndices[requestedIndex] {
-			allRequestedPresent = false
-			break
-		}
-	}
+// requestDirectSidecarsFromPeers tries to fetch missing data column sidecars from connected peers.
+// It searches through the available peers to identify those responsible for the requested columns,
+// and returns only after all columns have either been successfully retrieved or all candidate peers
+// have been exhausted.
+//
+// It returns a map from each root to its successfully retrieved sidecars.
+func requestDirectSidecarsFromPeers(
+	params DataColumnSidecarsParams,
+	roBlocks []blocks.ROBlock,
+	slotsWithCommitments map[primitives.Slot]bool,
+	storedIndicesByRoot map[[fieldparams.RootLength]byte]map[uint64]bool,
+	requestedIndices map[uint64]bool,
+	incompleteRoots map[[fieldparams.RootLength]byte]bool,
+) (map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn, error) {
+	start := time.Now()
 
-	if !allRequestedPresent {
-		return nil, nil
-	}
+	// Create a new random source for peer selection.
+	randomSource := rand.NewGenerator()
```

Comment:
Is this a todo down the road to change, or is it by design we usually do it this way?

Reply:
Hum no I don't plan to change that. What do you mean exactly? Do you think we should proceed differently?